data_mgr.py

```diff
@@ -0,0 +1,296 @@
+        natts = [
+            getattr(node, 'cycle_point', None),
+            getattr(node, 'namespace', [
+                n for n in [getattr(node, 'name', None)] if n is not None]),
+            getattr(node, 'submit_num', None),
+            getattr(node, 'state', None),
+        ]
+        if (
+                (not args.get('states') or node.state in args['states']) and
+                not (
+                    args.get('exstates') and
+                    node.state in args['exstates']) and
+                (args.get('mindepth', -1) < 0 or
+                    node.depth >= args['mindepth']) and
+                (args.get('maxdepth', -1) < 0 or
+                    node.depth <= args['maxdepth']) and
+                (not ids or self._node_atts_filter(natts, ids)) and
+                not (exids and self._node_atts_filter(natts, exids))):
+            return True
+        return False
+
+    @staticmethod
+    def _node_atts_filter(natts, items):
+        for point, name, submit_num, state in items:
+            if ((not point or fnmatchcase(natts[0], point)) and
+                    any(fnmatchcase(nn, name) for nn in natts[1]) and
+                    (not submit_num or fnmatchcase(natts[2], submit_num)) and
+                    (not state or natts[3] == state)):
+                return True
+        return False
+
+    def _workflow_filter(self, flow, ids=[], exids=[], args={}):
+        natts = [flow.workflow.owner, flow.workflow.name, flow.workflow.status]
+        if ((not ids or self._workflow_atts_filter(natts, ids)) and
+                not (exids and self._workflow_atts_filter(natts, exids))):
+            return True
+        return False
+
+    @staticmethod
+    def _workflow_atts_filter(natts, items):
+        for owner, name, status in set(items):
+            if ((not owner or fnmatchcase(natts[0], owner)) and
+                    (not name or fnmatchcase(natts[1], name)) and
+                    (not status or natts[2] == status)):
+                return True
+        return False
+
+    @staticmethod
+    def parse_workflow_id(item):
+        """Split workflow id argument to individual workflow attributes.
+        Args:
+            item (owner/workflow:status):
+                It's possible to traverse workflows,
+                defaults to UI Server owner, and ``*`` glob for workflow.
+
+        Returns:
+            (owner, name, status)
+
+        """
+        owner, workflow, status = (None, None, None)
+        if ':' in item:
+            head, status = item.rsplit(':', 1)
+        else:
+            head, status = (item, None)
+        if head.count('/'):
+            owner, workflow = head.split('/', 1)
+        else:
+            # more common to filter on workflow (with owner constant)
+            workflow = head
+        return (owner, workflow, status)
+
+    @staticmethod
+    def parse_node_id(item, node_type=None):
+        """Parse node id argument.
+        Args:
```

Comment:
The docstrings here are missing the arguments. I think the Google style docstrings do not support optional or tuples. So the `Items` entry below could be part of the documentation for `item` I think... or with PEP-484 tuple types can be specified too)

Reply:
I was just roughly following what Oliver did in the ZeroMQ (not that he did this), I need to read up on Google style docstrings...